types/src/v2/params.rs

```diff
@@ -89,40 +89,39 @@ impl Serialize for TwoPointZero {
 /// The data containing the params is a `Cow<&str>` and can either be a borrowed `&str` of JSON from an incoming
 /// [`super::request::JsonRpcRequest`] (which in turn borrows it from the input buffer that is shared between requests);
 /// or, it can be an owned `String`.
+///
+/// Regards empty `JSON array` and `JSON object` as no parameters provided.
 #[derive(Clone, Debug)]
-pub struct RpcParams<'a>(Option<Cow<'a, str>>);
+pub struct RpcParams<'a>(Cow<'a, str>);
 
 impl<'a> RpcParams<'a> {
 	/// Create params
 	pub fn new(raw: Option<&'a str>) -> Self {
-		Self(raw.map(Into::into))
+		Self(match raw.unwrap_or("").trim() {
+			"{}" | "[]" => "".into(),
```

Comment:
This just occurred to me: `{ }` or `[ ]` are valid JSON but will still run into an error here.

Reply:
https://github.com/paritytech/jsonrpsee/pull/446#discussion_r699070793

Reply:
lemme add a test for this,  but your point is that:

```rust
RpcParams::new("[]").parse::<JsonValue>() -> Err(InvalidParams)
RpcParams::new("{}").parse::<JsonValue>() -> Err(InvalidParams)
```

It should be treated as valid param right?
So are we back at having to different `RpcParams::sequence` and `RpcParams::sequence_ignore_empty`?

Reply:
Right, there is also another edge-case here when you have a single array/map as a parameter and call `sequence.next()` it will be rejected as an invalid param.

The simplest would just to provide `None` for `optional_next` in the proc macro code but it's not a good idea because parameters will ignored without telling the caller.